Improvments to the dateTimeConvert udf

npawar opened the pull request:
1) Remove redundant validations, keeping it only at the blockVal level, instead of for every record
2) Prevent the creation of new DateTimeFormatter for every record
3) Fix bug which prevented SDF format in input to work
4) Add support for : in sdf pattern
5) Remove code which replaced spaces in function parameters
6) Remove the assumption that time column data type is always long
7) Allow graceful handling of null values in input, instead of failing

kishoreg commented on pinot-core/src/main/java/com/linkedin/pinot/core/operator/transform/function/DateTimeConversionTransform.java:
```diff
@@ -65,28 +67,34 @@
-          DateTimeConversionTransformUtils.convertMillisToFormat(bucketedDateTimevalueMS,
-              outputFormat);
-      _output[i] = dateTimeValueConverted;
+    DataType valueType = input[0].getValueType();
```
lets use the eclipse stylesheet to format this code.

kishoreg commented on pinot-core/src/main/java/com/linkedin/pinot/core/operator/transform/function/DateTimeConversionTransform.java:
```diff
@@ -65,28 +67,34 @@
     String outputFormat = outputDateTimeFormat[0];
+    String[] outputDateTimeGranularity = input[3].getStringValuesSV();
     String outputGranularity = outputDateTimeGranularity[0];
+    DateTimeConversionEvaluator evaluator = new DateTimeConversionEvaluator(inputFormat, outputFormat, outputGranularity);
```
do we need to create one instance for every block?

kishoreg commented on pinot-core/src/main/java/com/linkedin/pinot/core/operator/transform/DateTimeConversionEvaluator.java:
```diff
@@ -0,0 +1,213 @@
+/**
+ * Converts the time column value based on the inputFormat, outputFormat and outputGranularity
+ */
+public class DateTimeConversionEvaluator {
```
why Evaluator instead of converter/transformer/formatter

kishoreg commented on pinot-core/src/main/java/com/linkedin/pinot/core/operator/transform/DateTimeConversionEvaluator.java:
```diff
@@ -0,0 +1,213 @@
+   * @param inputValue
+   * @return
+   */
+  public long transformDateTime(Object inputValue) {
```
these can be optimized further. We should set up everything in the constructor so that the processing for each value is low.

kishoreg commented on pinot-core/src/main/java/com/linkedin/pinot/core/operator/transform/DateTimeConversionEvaluator.java:
```diff
@@ -0,0 +1,213 @@
+  private Long convertFromInputFormatToMillis(Object dateTimeValue) {
+
+    Long timeColumnValueMS = 0L;
+    if (dateTimeValue != null) {
```
instead of checking this on a per record basis, setup epochToSDFConvertor and SDFToEpochConverter etc and pick the right converter in the beginning.

npawar commented:
Addressed the comments @kishoreg . PLease check

kishoreg commented:
I forgot to click on submit for previous comments. I will review the latest changes now

kishoreg commented on pinot-common/src/main/java/com/linkedin/pinot/common/utils/time/DateTimeFieldSpecUtils.java:
```diff
@@ -28,17 +28,17 @@
 
 public class DateTimeFieldSpecUtils {
```
It would be nice to have a class TimeSpecFormat that takes in the format and provides all these getters.

kishoreg commented on pinot-core/src/main/java/com/linkedin/pinot/core/operator/transform/DateTimeConvertor.java:
```diff
@@ -0,0 +1,213 @@
+  /**
+   * Time unit enum with range from MILLISECONDS to YEARS
+   */
+  public enum DateTimeTransformUnit {
```
Move this to the DateTimeFormat class (current datatimespecutils class)

kishoreg commented on pinot-core/src/main/java/com/linkedin/pinot/core/operator/transform/DateTimeConvertor.java:
```diff
@@ -0,0 +1,213 @@
+public abstract class DateTimeConvertor {
+
+  private static final DateTime EPOCH_START_DATE = new DateTime(0L, DateTimeZone.UTC);
+  private static final Chronology CHRONOLOGY = DateTimeUtils.getInstantChronology(EPOCH_START_DATE);
```
rename CHRONOLOGY to EPOCH_START_CHRONOLOGY

npawar commented:
Moved the DateTimeConvertor classes to another package(com.linkedin.pinot.common.datetime.convertor)
Also introduced classes DateTimeFormatSpec and DateTimeGranularitySpec to hold the getters, constructors and validators, hence getting rid of DateTimeFieldSpecUtils
